### PR TITLE
Fix slice bug

### PR DESCRIPTION
**Pull request by Flowingsun007**
修复tensor切片的bug，支持对size=1的dim自动squeeze

### CODE REVIEW

File: oneflow/python/test/modules/test_slice.py [hunk 1]
```diff
@@ -42,6 +42,20 @@ def _test_slice_4_dim(test_case, device):
     test_case.assertTrue(np.array_equal(y.numpy(), np_out))
 
 
+def _test_slice_special_case(test_case, device):
```
Comment (BBuf) [inline oneflow/python/test/modules/test_slice.py:45]: 测试样例把xuzhiqiu那个例子也加上吧。

Comment (BBuf) [reply oneflow/python/test/modules/test_slice.py:45]:
```python
>>> import torch
>>> a = torch.Tensor([[[1, 2], [3, 4]], [[5, 6], [7, 8]]])
>>> a[0].shape
torch.Size([2, 2])
>>> a[0, :, :].shape
torch.Size([2, 2])
>>> a[0][1][1]
tensor(4.)
>>> import oneflow as flow
>>> a = flow.Tensor([[[1, 2], [3, 4]], [[5, 6], [7, 8]]])
>>> a[0].shape
flow.Size([1, 2, 2])
>>> a[0, :, :].shape
flow.Size([1, 2, 2])
>>> a[0][1][1]
```

File: oneflow/python/framework/tensor.py [hunk 2]
```diff
@@ -404,7 +404,7 @@ def __getitem__(self, key):
         # TODO: support inplace __getitem__
         start, stop, step, _ = self._get_slice_obj(key)
         res = flow.experimental.slice(self, list(zip(start, stop, step)))
-        return res
+        return res.squeeze()
```
Comment (daquexian) [inline oneflow/python/framework/tensor.py:407]: 不能直接这样，会让 a[1:2] 这种情况也被 squeeze，应该先根据 key 统计一下哪些维度的下标是单纯的数字而不是 `1:2` 这种

File: oneflow/python/nn/modules/squeeze.py [hunk 3]
```diff
@@ -67,9 +70,21 @@ def squeeze_op(input, dim: Optional[Sequence[int]] = None):
...
+    squeeze_dim = []
+    if dim is None:
+        for i in range(input.ndim):
+            if input.size(i) == 1:
+                squeeze_dim.append(i)
+    else:
+        if type(dim) == int:
+            if input.size(dim) == 1:
+                squeeze_dim.append(dim)
+        else:
```
Comment (daquexian) [inline oneflow/python/nn/modules/squeeze.py:82]: 先得到一个 list 类型的 dim，就可以统一处理，不用重复三遍了

File: oneflow/python/framework/tensor.py (hunk 2 above)
Comment (Flowingsun007) [reply oneflow/python/framework/tensor.py:407]: 有道理

File: oneflow/python/nn/modules/squeeze.py (hunk 3 above)
Comment (Flowingsun007) [reply oneflow/python/nn/modules/squeeze.py:82]: 好的

Comment (Flowingsun007) [reply oneflow/python/nn/modules/squeeze.py]: 已修改

File: oneflow/python/test/modules/test_slice.py (hunk 1 above)
Comment (Flowingsun007) [reply oneflow/python/test/modules/test_slice.py:45]: 好的

File: oneflow/python/nn/modules/squeeze.py [hunk 4]
```diff
@@ -69,7 +72,17 @@ def squeeze_op(input, dim: Optional[Sequence[int]] = None):
     """
     if type(dim) == int:
```
Comment (daquexian) [inline oneflow/python/nn/modules/squeeze.py:73]:
```suggestion
    if isinstance(dim, int):
```

File: oneflow/python/nn/modules/squeeze.py [hunk 5]
```diff
@@ -69,7 +72,17 @@ def squeeze_op(input, dim: Optional[Sequence[int]] = None):
...
+    squeeze_dim = []
+    if dim is None:
+        for i in range(input.ndim):
+            if input.size(i) == 1:
+                squeeze_dim.append(i)
+    else:
+        for i in dim:
+            if input.size(i) == 1:
+                squeeze_dim.append(i)
+    return Squeeze(dim=squeeze_dim)(input)
```
Comment (daquexian) [inline oneflow/python/nn/modules/squeeze.py:85]:
```suggestion
    if dim is None:
        dim = range(input.ndim)
 
    for i in dim:
        if input.size(i) == 1:
            squeeze_dim.append(i)
    return Squeeze(dim=squeeze_dim)(input)
```

File: oneflow/python/nn/modules/squeeze.py [hunk 6]
```diff
@@ -67,9 +70,17 @@ def squeeze_op(input, dim: Optional[Sequence[int]] = None):
...
 
     """
-    if type(dim) == int:
+    if isinstance(dim, int):
         dim = [dim]
-    return Squeeze(dim=dim)(input)
+
+    squeeze_dim = []
+    if dim is None:
+        dim = range(input.ndim)
```
Comment (daquexian) [inline oneflow/python/nn/modules/squeeze.py:78]: 77-78 行和上面的 73-74 行写在一起吧

File: oneflow/python/nn/modules/squeeze.py [hunk 7]
```diff
@@ -67,9 +70,17 @@ def squeeze_op(input, dim: Optional[Sequence[int]] = None):
...
+
+    for i in dim:
+        if input.size(i) == 1:
+            squeeze_dim.append(i)
+    return Squeeze(dim=squeeze_dim)(input)
```
Comment (daquexian) [inline oneflow/python/nn/modules/squeeze.py:83]:
```suggestion
    dim = list(filter(lambda i: input.size(i) == 1, dim))
    return Squeeze(dim=dim)(input)
```

squeeze_dim 可以删掉了

File: oneflow/python/test/modules/test_slice.py [hunk 8]
```diff
@@ -42,6 +42,26 @@ def _test_slice_4_dim(test_case, device):
...
+    x = flow.Tensor(np_arr, device=flow.device(device))
+    of_out = x[0, :, :]
+    np_out = np_arr[0, :, :]
+    test_case.assertTrue(np.array_equal(of_out.numpy(), np_out))
+
+    np_arr = np.random.randn(2, 3, 4, 5).astype(np.float32)
+    x = flow.Tensor(np_arr, device=flow.device(device))
+    of_out = x[0, :, :, :]
+    np_out = np_arr[0, :, :, :]
+    test_case.assertTrue(np.array_equal(of_out.numpy(), np_out))
```
Comment (daquexian) [inline oneflow/python/test/modules/test_slice.py:62]: 再专门测一下 `x[1:2]` 这种 slicing 吧

File: oneflow/python/test/modules/test_slice.py [hunk 9]
```diff
@@ -42,6 +42,26 @@ def _test_slice_4_dim(test_case, device):
     test_case.assertTrue(np.array_equal(y.numpy(), np_out))
 
 
+def _test_slice_special_case(test_case, device):
```
Comment (daquexian) [inline oneflow/python/test/modules/test_slice.py:45]: 这个名字应该更有描述性

Comment (Flowingsun007) [reply oneflow/python/test/modules/test_slice.py:45]: 这个名字怎么起呢？

File: oneflow/python/test/modules/test_slice.py (hunk 8 above)
Comment (Flowingsun007) [reply oneflow/python/test/modules/test_slice.py:68]: 好的，已增加test case

File: oneflow/python/nn/modules/squeeze.py (hunk 7 above)
Comment (Flowingsun007) [reply oneflow/python/nn/modules/squeeze.py]: 好的

File: oneflow/python/nn/modules/squeeze.py (hunk 6 above)
Comment (Flowingsun007) [reply oneflow/python/nn/modules/squeeze.py:76]: 已修改

**Review by daquexian — APPROVED** [review]

**Review by BBuf — APPROVED** [review]
